pkg/destroy/alibabacloud/alibabacloud.go

```diff
@@ -0,0 +1,1345 @@
+	o.Logger.Debugf("Start to detach RAM policy %q with %q", policyName, principalName)
+	request := resourcemanager.CreateDetachPolicyRequest()
+	request.Scheme = "https"
+	request.PolicyName = policyName
+	request.PolicyType = policyType
+	request.PrincipalName = principalName
+	request.PrincipalType = principalType
+	request.ResourceGroupId = resourceGroupID
+	_, err = o.rmanagerClient.DetachPolicy(request)
+	return
+}
+
+func (o *ClusterUninstaller) deletePrivateZones() (err error) {
+	clusterDomain := o.ClusterDomain
+	o.Logger.Debug("Start to search private zones")
+	zones, err := o.listPrivateZone(clusterDomain)
+	if err != nil {
+		return err
+	}
+	if len(zones) == 0 {
+		return nil
+	}
+	if len(zones) > 1 {
+		return errors.Wrap(err, fmt.Sprintf("matched to multiple private zones by clustedomain %q", clusterDomain))
+	}
+
+	zoneID := zones[0].ZoneId
+	err = o.bindZoneVpc(zoneID)
+	if err != nil {
+		return err
+	}
+
+	// Wait for unbind vpc to complete
+	err = wait.Poll(
+		1*time.Second,
+		1*time.Minute,
+		func() (bool, error) {
+			zones, err := o.listPrivateZone(clusterDomain)
+			if err != nil {
+				return false, err
+			}
+
+			if len(zones[0].Vpcs.Vpc) == 0 {
+				return true, nil
+			}
+			return false, nil
+		},
+	)
+	if err != nil {
+		return
+	}
+
+	o.Logger.Debug("Start to delete private zones")
+	// Delete a private zone does not require delete the record in advance
+	err = o.deletePrivateZone(zoneID)
+	if err != nil {
+		return err
+	}
+
+	// Wait for deletion private zone to complete
+	err = wait.Poll(
+		1*time.Second,
+		1*time.Minute,
+		func() (bool, error) {
+			zones, err := o.listPrivateZone(clusterDomain)
+			if err != nil {
+				return false, err
+			}
+
+			if len(zones) == 0 {
+				return true, nil
+			}
+			return false, nil
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *ClusterUninstaller) deletePrivateZone(zoneID string) (err error) {
+	o.Logger.Debugf("Start to delete private zone %q", zoneID)
+	request := pvtz.CreateDeleteZoneRequest()
+	request.ZoneId = zoneID
+	_, err = o.pvtzClient.DeleteZone(request)
+	return
+}
+
+func (o *ClusterUninstaller) bindZoneVpc(zoneID string) (err error) {
+	o.Logger.Debugf("Start to unbind/bind private zone %q with vpc", zoneID)
+	request := pvtz.CreateBindZoneVpcRequest()
+	request.ZoneId = zoneID
+	_, err = o.pvtzClient.BindZoneVpc(request)
+	return
+}
+
+func (o *ClusterUninstaller) listPrivateZone(clusterDomain string) ([]pvtz.Zone, error) {
+	request := pvtz.CreateDescribeZonesRequest()
+	request.Lang = "en"
+	request.Keyword = clusterDomain
+	request.SearchMode = "EXACT"
+
+	response, err := o.pvtzClient.DescribeZones(request)
+	if err != nil {
+		return nil, err
+	}
+	return response.Zones.Zone, nil
+}
+
+func (o *ClusterUninstaller) deleteDNSRecords() (err error) {
+	o.Logger.Debug("Start to search DNS records")
+
+	baseDomain := strings.Join(strings.Split(o.ClusterDomain, ".")[1:], ".")
```

Comment:
Is this the fix for the fuzzy domain search error? If so can you explain how this solves the problem? Also it would be nice to have a comment in the code  for why we need the string manipulation

Reply:
Aha~, about azure stack commit, it is a rebase mistake, I will updata it later.
The `LIKE` is default search mode on  DescribeDomains api. In this mode, if you search `openshift.com`,  can match `xx.openshift.com` and `openshift.com.xx`, but using `EXACT` mode, only return `openshift.com`.

> Also it would be nice to have a comment in the code for why we need the string manipulation

I need base domain to search domain and delete it records, but I can not get base domain, so I use cluster domin to splite by `.` (the format of cluster domain is `<cluster name>.<base domain>`).

I will add comment here.